File: bmath/unit/Velocity.go
package unit

import "fmt"

//VelocityMPS is the value indicating that velocity value is expressed in meters per second
const VelocityMPS byte = 60

//VelocityKMH is the value indicating that velocity value is expressed in kilometers per hour
const VelocityKMH byte = 61

//VelocityFPS is the value indicating that velocity value is expressed in feet per second
const VelocityFPS byte = 62

//VelocityMPH is the value indicating that velocity value is expressed in miles per hour
const VelocityMPH byte = 63

//VelocityKT is the value indicating that velocity value is expressed in knots
const VelocityKT byte = 64

func velocityToDefault(value float64, units byte) (float64, error) {
	switch units {
	case VelocityMPS:
		return value, nil
	case VelocityKMH:
		return value / 3.6, nil
	case VelocityFPS:
		return value / 3.2808399, nil
	case VelocityMPH:
		return value / 2.23693629, nil
	case VelocityKT:
		return value / 1.94384449, nil
	default:
		return 0, fmt.Errorf("Velocity: unit %d is not supported", units)
	}
}

func velocityFromDefault(value float64, units byte) (float64, error) {
	switch units {
	case VelocityMPS:
		return value, nil
	case VelocityKMH:
		return value * 3.6, nil
	case VelocityFPS:
		return value * 3.2808399, nil
	case VelocityMPH:
		return value * 2.23693629, nil
	case VelocityKT:
		return value * 1.94384449, nil
	default:
		return 0, fmt.Errorf("Velocity: unit %d is not supported", units)
	}
}

//Velocity struct keeps velocity or speed values
type Velocity struct {
	value        float64
	defaultUnits byte
}

//CreateVelocity creates a velocity value.
//
//units are measurement unit and may be any value from
//unit.Velocity_* constants.
func CreateVelocity(value float64, units byte) (Velocity, error) {
	v, err := velocityToDefault(value, units)
	if err != nil {
		return Velocity{}, err
	}
	return Velocity{value: v, defaultUnits: units}, nil

}

//MustCreateVelocity creates the velocity value but panics instead of returned a error
func MustCreateVelocity(value float64, units byte) Velocity {
	v, err := CreateVelocity(value, units)
	if err != nil {
		panic(err)
	}
	return v
}

//Value returns the value of the velocity in the specified units.
//
//units are measurement unit and may be any value from
//unit.Velocity_* constants.
//
//The method returns a error in case the unit is
//not supported.
func (v Velocity) Value(units byte) (float64, error) {
	return velocityFromDefault(v.value, units)
}

//Convert converts the value into the specified units.
//
//units are measurement unit and may be any value from
//unit.Velocity_* constants.
func (v Velocity) Convert(units byte) Velocity {
	return Velocity{value: v.value, defaultUnits: units}
}

//In converts the value in the specified units.
//Returns 0 if unit conversion is not possible.
func (v Velocity) In(units byte) float64 {
	x, e := velocityFromDefault(v.value, units)
	if e != nil {
		return 0
	}
	return x

}

func (v Velocity) String() string {
	x, e := velocityFromDefault(v.value, v.defaultUnits)
	if e != nil {
		return "!error: default units aren't correct"
	}
	var unitName, format string
	var accuracy int
	switch v.defaultUnits {
	case VelocityMPS:
		unitName = "m/s"
		accuracy = 0
	case VelocityKMH:
		unitName = "km/h"
		accuracy = 1
	case VelocityFPS:
		unitName = "ft/s"
		accuracy = 1
	case VelocityMPH:
		unitName = "mph"
		accuracy = 1
	case VelocityKT:
		unitName = "kt"
		accuracy = 1
	default:
		unitName = "?"
		accuracy = 6
	}
	format = fmt.Sprintf("%%.%df%%s", accuracy)
	return fmt.Sprintf(format, x, unitName)

}

//Units return the units in which the value is measured
func (v Velocity) Units() byte {
	return v.defaultUnits
}

File: bmath/unit/velocity.py
from bmath.unit.types import UnitsConvertor


class VelocityConvertor(UnitConvertor):
    _unit_type = 'velocity'

    # the value indicating that weight value is expressed in some unit
    VelocityMPS = 70
    VelocityKMH = 71
    VelocityFPS = 72
    VelocityMPH = 73
    VelocityKT = 74

    _units = {
        VelocityMPS: {'name': 'm/s', 'accuracy': 0,
                      'to': lambda v: v,
                      'from': lambda v: v},
        VelocityKMH: {'name': 'km/h', 'accuracy': 1,
                      'to': lambda v: v / 3.6,
                      'from': lambda v: v * 3.6},
        VelocityFPS: {'name': 'ft/s', 'accuracy': 1,
                      'to': lambda v: v / 3.2808399,
                      'from': lambda v: v * 3.2808399},
        VelocityMPH: {'name': 'mph', 'accuracy': 1,
                      'to': lambda v: v / 2.23693629,
                      'from': lambda v: v * 2.23693629},
        VelocityKT: {'name': 'kt', 'accuracy': 1,
                     'to': lambda v: v / 1.94384449,
                     'from': lambda v: v * 1.94384449},
    }


class Velocity(object):
    """ Velocity object keeps velocity or speed values """

    def __init__(self, value: float, units: int):
        f"""
        Creates a velocity value
        :param value: velocity value
        :param units: TemperatureUnits.consts
        """
        v, err = VelocityConvertor().to_default(value, units)
        if err:
            self._value = None
            self._defaultUnits = None
        else:
            self._value = v
            self._default_units = units
        self.error = err

    def __str__(self):
        v, err = VelocityConvertor().from_default(self.v, self.default_units)
        if err:
            return f'Velocity: unit {self.default_units} is not supported'
        multiplier, name, accuracy = VelocityConvertor()(self.default_units)
        return f'{round(v, accuracy)} {name}'

    @staticmethod
    def must_create(value: float, units: int) -> float:
        """
        Returns the velocity value but panics instead of return error
        :param value: velocity value
        :param units: TemperatureUnits.consts
        :return: None
        """
        v, err = VelocityConvertor().to_default(value, units)
        if err:
            raise ValueError(f'Velocity: unit {units} is not supported')
        else:
            return v

    @staticmethod
    def value(velocity: 'Velocity', units: int) -> [float, Exception]:
        """
        :param velocity: Velocity
        :param units: TemperatureUnits.consts
        :return: Value of the velocity in the specified units
        """
        return VelocityConvertor().from_default(velocity.v, units)

    @staticmethod
    def convert(velocity: 'Velocity', units: int) -> 'Velocity':
        """
        Returns the value into the specified units
        :param velocity: Velocity
        :param units: TemperatureUnits.consts
        :return: Velocity object in the specified units
        """
        return Velocity(velocity.v, units)

    @staticmethod
    def convert_in(velocity: 'Velocity', units: int) -> [float, Exception]:
        """
        Converts the value in the specified units.
        Returns 0 if unit conversion is not possible.
        :param velocity: Velocity
        :param units: TemperatureUnits.consts
        :return: float
        """
        v, err = VelocityConvertor().from_default(velocity.v, units)
        if err:
            return 0
        return v

    @property
    def v(self):
        return self._value

    @property
    def default_units(self):
        return self._default_units


if __name__ == '__main__':
    pass
